Gems/Atom/RHI/Code/Source/RHI/Object.cpp

```diff
@@ -8,11 +8,18 @@
 #include <Atom/RHI/Object.h>
 #include <AzCore/Module/Environment.h>
 #include <AzCore/std/parallel/atomic.h>
+#include <AzCore/std/parallel/scoped_lock.h>
 
 namespace AZ::RHI
 {
+    //TODO see comment in Object.h
+#if RHI_OBJECT_STATIC_MUTEX
+    AZStd::recursive_mutex Object::s_nameMutex;
+#endif
+
     void Object::SetName(const Name& name)
     {
+        AZStd::scoped_lock<decltype(m_nameMutex)> lock(m_nameMutex);
```

Comment:
I am really surprised that this was not caught earlier. What was deadlocking in this case? The load level increase and memory increase is not good here, especially for mobile :(

Reply:
There are a few reasons why this was not caught as often before.
Before, the `m_databaseMutex` was being locked for the whole creation of InstanceData<type>. That changed now, because due to AsyncUploadQueue, `m_databaseMutex` is unlocked, which opens for more opportunities of MeshDrawPacket::DoUpdate() to run concurrently.